### ISSUE TITLE
How to access the info after "successful" ID Provider login?

### ISSUE THREAD

**Issue by lorezyra** [heiglandreas/HybridAuth#21]
I've updated my test / home page to output the data as you specified on the /README.md file. However, this seems to return ```NULL```. Can you provide suggestions as to why the container would return nothing?

/config/autoload/vendor.OrgHeiglHybridAuth.global.php:
```PHP
namespace OrgHeiglHybridAuth;


return array( 'OrgHeiglHybridAuth' => array(
    'hybrid_auth' => array(
       // 'base_url' => 'http://example.com/auth/backend',
        'providers' => array(
			"Facebook" => array (
				"enabled" => true,
				//"keys"    => array ( "id" => "...", "secret" => "..." ),
				"keys"    => array ( "id" => "...", "secret" => "..." ), // Lins­p­i­r­a­.com - Test1
				"trustForwarded" => false
			),

			"Twitter" => array (
				"enabled" => true,
				"keys"    => array ( "key" => "...", "secret" => "..." )
//				"keys"    => array ( "key" => "...", "secret" => "..." )
			),
        ),
        'debug_mode' => true,
        // For some reason Hybrid_Auth doesn't create file in a specified folder, be sure that it exists
		'debug_file' => ROOT_PATH . "data/log/hybridAuth.log"
    ),
    'session_name' => 'Linspira_hybridauth',
));
```

indexController.php:
```PHP
namespace Common\Mvc\Controller;

use Hybridauth\Hybridauth;
use Common\Mvc\Controller\BaseController;
use Zend\View\Model\ViewModel;

class IndexController extends BaseController
{
	public function indexAction() {
		$viewData = array();
		
		if($this->isUserAuthenticated()){
			return $this->redirect()->toRoute('feed');
		}
		
		$service = $this->getServiceLocator();
		$service->get('translator')->setLocale('en_US'); //TODO: this must be set to user's setting
		
		try{
		// Need this block to autoload Hybrid_Auth dependencies to unserialize object stored in session
		$config = $service->get('Config');
		$config = $config['OrgHeiglHybridAuth'];
		$hybridAuth = new Hybridauth($config['hybrid_auth']);
//		var_dump($hybridAuth);
		
		$container = new \Zend\Session\Container('Linspira_hybridauth');
//		var_dump($container);
		} catch (Exception $e) {}


//		$viewData['config'] = $config;
		$viewData['hybridAuth'] = $hybridAuth;
		$viewData['container'] = $container;
		return new ViewModel($viewData);
	}
}
```
index.phtml:
```PHP

var_dump($this->container);


if ( $this->container->offsetExists('authenticated')) {

	/** @var OrgHeiglHybridAuth\UserInterface $user */
	$user = $this->container->offsetGet('user');
	var_dump($user);

	if( $user !== NULL) {
		echo $user->getName(); // The name of the logged in user
		echo $user->getUID();  // The internal UID of the used service
		echo $user->getMail(); // The mail-address the service provides
		echo $user->getLanguage(); // The language the service provides for the user
		$service = $container->offsetGet('backend');
		echo $service->id; // Should print out the Name of the service provider.
	}
} else {
	echo 'Not signed in via SSO...';
}
```

I must be doing something wrong... Can you point me in the right direction?

**Comment by heiglandreas** [comment]
Ah, yes. I think the issue is the ```use Hybridauth\Hybridauth```-line in the IndexController. Replace that one with ```use OrgHeiglHybridAuth\Hybridauth```. That should do the trick.

Currently you are using the Hybridauth-Library directly which (IIRC) doesn't understand the config-file that is made for the OrgHeiglHybridAuth-Lib :wink:

**Comment by lorezyra** [comment]
I changed to use ```OrgHeiglHybridAuth\Hybridauth``` and it errors out. Checking the stack trace, I see the following errors upstream:
```
URI: /authenticate/login/Facebook/aG9tZQ==
Date: 2015-05-03 19:20:04
File: /var/www/vendor/ZF2/library/Zend/ServiceManager/ServiceManager.php

An exception was raised while creating "OrgHeiglHybridAuth\Controller\IndexController"; no instance returned
An exception was raised while creating "OrgHeiglHybridAuthBackend"; no instance returned
```

I also see this error from time to time:
```
File: /var/www/vendor/hybridauth/hybridauth/src/Hybridauth/Storage/Session.php
Message: Serialization of 'Closure' is not allowed

```

Is there no better way to handle provider errors?

**Comment by heiglandreas** [comment]
The two Exceptions from above are thrown by the ServiceManager of ZF2. Do you still have the ```backend_url``` commented in the config from above? It has to be present and should contain the URL of the page you want to redirect the user after a successfull login.

**Comment by lorezyra** [comment]
Oh, so the ```base_url``` config is **not** optional?

For some reason, I thought OrgHeiglHybridAuth would handle the return url and I would just grab the container data after returning to "home."

In that case, I should have it route to my user controller and parse the data before redirecting to the user's new home page.

**Comment by heiglandreas** [comment]
As, depending on your app, the redirect-page can be different from the home-url that config-parameter either has to be given or to be overwritten depending on your needs. As it's the URL that will be called from the AuthProvider after a successfull authentication it has to be given somewhere. 

You are right, in that I could have set the 'home'-route as default, but there might be no route 'home' :frowning:. Hence the config-param.

**Comment by lorezyra** [comment]
Okay... :stuck_out_tongue_closed_eyes:  I still have some hair left.

I have *not* been successful over the last week in getting this to work. I admit that I'm not the greatest programmer that breathed on code... However, I decided to start over from the top.

I removed both OrgHeiglHybridAuth & Hybridauth submodules after I realized Hybridauth was a combination of the old and new code bases. After deleting and re-adding OrgHeiglHybridAuth submodule, I manually downloaded the Hybridauth library (v3+). Then, I created the classMap file. The ```createAutoloadSupport.php``` didn't work for me, so I had to change it a bit to get the results I required.

```PHP

$VENDOR_FOLDER = realpath(getenv('VENDOR_FOLDER'));

if ( $VENDOR_FOLDER !== NULL ) {
    $currentFolder = dirname(__FILE__);
    $vendorPos = strpos($currentFolder, '/vendor/');
    if ($vendorPos !== false) {
        $VENDOR_FOLDER = realpath(substr($currentFolder, 0, $vendorPos + 7));
    }
}

//echo "this file path: " . __FILE__ . "\n\n";
//echo "VENDOR_FOLDER: " . $VENDOR_FOLDER . "\n\n";

$ORG_HEIGL_FOLDER =   dirname( dirname( __FILE__ ));
echo "ORG_HEIGL_FOLDER: " . $ORG_HEIGL_FOLDER . "\n";

//if ( !$VENDOR_FOLDER ) {
//   $VENDOR_FOLDER = realpath( dirname(__FILE__) . '/../vendor');
//}

if ( !$VENDOR_FOLDER ) {
	die ('No vendor-Folder found!');
}

echo "VENDOR_FOLDER: " . $VENDOR_FOLDER . "\n\n";

$cmdStr = NULL;

$cmd[] = 'cd "' . $VENDOR_FOLDER . '/hybridauth/src' . '"'; //MAKE SURE YOU HAVE THE 3.0.X VERSION !!!
$cmd[] = $VENDOR_FOLDER . '/ZF2/bin/classmap_generator.php -o  ' . $ORG_HEIGL_FOLDER . '/autoload_classmap.php';
$cmd[] = 'cd "' . dirname(__FILE__)  . '/../" ';
$cmd[] = $VENDOR_FOLDER . '/ZF2/bin/classmap_generator.php -a -o  ' . $ORG_HEIGL_FOLDER . '/autoload_classmap.php';

//var_dump ($cmd); 

$cmdStr = implode(' ; ', $cmd);

echo ($cmdStr);
exec( $cmdStr );
```

So now, when I click on my SSO link for FB, it goes to ```http://Linspira.com/authenticate/login/Facebook/aG9tZQ== ``` and I get prompted by FB to authorize. Then, it processes the return_uri to return to my home page. I can see that both modules are working up to this point.

However, my trouble comes when I try to capture the data returned by FB. The example provided on this project's home page doesn't work out-of-the-box. When I attempt to search for the class/method ```Hybridauth``` from this code (IndexController.php):
```PHP
namespace Common\Mvc\Controller;

use Common\Mvc\Controller\BaseController;
use Zend\View\Model\ViewModel;
use OrgHeiglHybridAuth\Module;
use Hybridauth\Hybridauth;

class IndexController extends BaseController
{
	public function indexAction() {
		$viewData = array();
		
		if($this->isUserAuthenticated()){
			return $this->redirect()->toRoute('feed');
		}

if ( !class_exists('Hybridauth\Hybridauth') ) {
    echo "Could not find Hybridauth class!\n";
} else {
//if ( !class_exists('OrgHeiglHybridAuth\Hybridauth') ) {
//    echo "Could not find Hybridauth class!\n";
//} else {


//	try{
		// Need this block to autoload Hybrid_Auth dependencies to unserialize object stored in session
		$config = $this->getServiceLocator()->get('Config');
		$config = $config['OrgHeiglHybridAuth'];
//	var_dump($config);
		$hybridAuth = new Hybridauth($config['hybrid_auth']);
//		$hybridAuth = new \OrgHeiglHybridAuth\Hybridauth($config['hybrid_auth']);
//var_dump($hybridAuth);
		
		$container = new \Zend\Session\Container('Linspira_hybridauth');
var_dump($container);
//	} catch (Exception $e) {}


//		$viewData['config'] = $config;
		$viewData['hybridAuth'] = $hybridAuth;
		$viewData['container'] = $container;
		return new ViewModel($viewData);
	}
    
}
//}
}

```

I fail to find the class anywhere in the OrgHeiglHybridAuth library... (I do, however, find the class mentioned in the view-helper... but that isn't what initializes the code base.) Even the classmap file doesn't show a match for "OrgHeiglHybridAuth\Hybridauth". I can see by cycling through the loaded Modules, that both modules are in fact loaded.

I must be missing something (aside from sleep).

**Comment by lorezyra** [comment]
Also, I might suggest adding the static ```autoload_classmap.php``` file directly into the project. This is what the other (https://github.com/SocalNick/ScnSocialAuth) ZF2 lib does.

Here's the output I got when I ran the modified script:
:memo: /vendor/OrgHeiglHybridAuth/autoload_classmap.php
```PHP
<?php
// Generated by ZF2's ./bin/classmap_generator.php
return array(
    'Hybridauth\Adapter\AbstractAdapter'                          => __DIR__ . '/../hybridauth/src/Adapter/AbstractAdapter.php',
    'Hybridauth\Adapter\AdapterInterface'                         => __DIR__ . '/../hybridauth/src/Adapter/AdapterInterface.php',
    'Hybridauth\Adapter\OAuth1'                                   => __DIR__ . '/../hybridauth/src/Adapter/OAuth1.php',
    'Hybridauth\Adapter\OAuth2'                                   => __DIR__ . '/../hybridauth/src/Adapter/OAuth2.php',
    'Hybridauth\Adapter\OpenID'                                   => __DIR__ . '/../hybridauth/src/Adapter/OpenID.php',
    'Hybridauth\Data\Collection'                                  => __DIR__ . '/../hybridauth/src/Data/Collection.php',
    'Hybridauth\Data\Parser'                                      => __DIR__ . '/../hybridauth/src/Data/Parser.php',
    'Hybridauth\Deprecated\DeprecatedAdapterTrait'                => __DIR__ . '/../hybridauth/src/Deprecated/DeprecatedAdapterTrait.php',
    'Hybridauth\Deprecated\DeprecatedHybridauthTrait'             => __DIR__ . '/../hybridauth/src/Deprecated/DeprecatedHybridauthTrait.php',
    'Hybridauth\Deprecated\DeprecatedStorageTrait'                => __DIR__ . '/../hybridauth/src/Deprecated/DeprecatedStorageTrait.php',
    'Hybridauth\Exception\AuthorizationDeniedException'           => __DIR__ . '/../hybridauth/src/Exception/AuthorizationDeniedException.php',
    'Hybridauth\Exception\BadMethodCallException'                 => __DIR__ . '/../hybridauth/src/Exception/BadMethodCallException.php',
    'Hybridauth\Exception\Exception'                              => __DIR__ . '/../hybridauth/src/Exception/Exception.php',
    'Hybridauth\Exception\ExceptionInterface'                     => __DIR__ . '/../hybridauth/src/Exception/ExceptionInterface.php',
    'Hybridauth\Exception\HttpClientFailureException'             => __DIR__ . '/../hybridauth/src/Exception/HttpClientFailureException.php',
    'Hybridauth\Exception\HttpRequestFailedException'             => __DIR__ . '/../hybridauth/src/Exception/HttpRequestFailedException.php',
    'Hybridauth\Exception\InvalidAccessTokenException'            => __DIR__ . '/../hybridauth/src/Exception/InvalidAccessTokenException.php',
    'Hybridauth\Exception\InvalidApplicationCredentialsException' => __DIR__ . '/../hybridauth/src/Exception/InvalidApplicationCredentialsException.php',
    'Hybridauth\Exception\InvalidArgumentException'               => __DIR__ . '/../hybridauth/src/Exception/InvalidArgumentException.php',
    'Hybridauth\Exception\InvalidAuthorizationCodeException'      => __DIR__ . '/../hybridauth/src/Exception/InvalidAuthorizationCodeException.php',
    'Hybridauth\Exception\InvalidAuthorizationStateException'     => __DIR__ . '/../hybridauth/src/Exception/InvalidAuthorizationStateException.php',
    'Hybridauth\Exception\InvalidOauthTokenException'             => __DIR__ . '/../hybridauth/src/Exception/InvalidOauthTokenException.php',
    'Hybridauth\Exception\InvalidOpenidIdentifierException'       => __DIR__ . '/../hybridauth/src/Exception/InvalidOpenidIdentifierException.php',
    'Hybridauth\Exception\RuntimeException'                       => __DIR__ . '/../hybridauth/src/Exception/RuntimeException.php',
    'Hybridauth\Exception\UnexpectedValueException'               => __DIR__ . '/../hybridauth/src/Exception/UnexpectedValueException.php',
    'Hybridauth\Exception\UnsupportedFeatureException'            => __DIR__ . '/../hybridauth/src/Exception/UnsupportedFeatureException.php',
    'Hybridauth\HttpClient\Curl'                                  => __DIR__ . '/../hybridauth/src/HttpClient/Curl.php',
    'Hybridauth\HttpClient\HttpClientInterface'                   => __DIR__ . '/../hybridauth/src/HttpClient/HttpClientInterface.php',
    'Hybridauth\HttpClient\Util'                                  => __DIR__ . '/../hybridauth/src/HttpClient/Util.php',
    'Hybridauth\Hybridauth'                                       => __DIR__ . '/../hybridauth/src/Hybridauth.php',
    'Hybridauth\Logger\Logger'                                    => __DIR__ . '/../hybridauth/src/Logger/Logger.php',
    'Hybridauth\Logger\LoggerInterface'                           => __DIR__ . '/../hybridauth/src/Logger/LoggerInterface.php',
    'Hybridauth\Provider\AOL'                                     => __DIR__ . '/../hybridauth/src/Provider/AOL.php',
    'Hybridauth\Provider\Disqus'                                  => __DIR__ . '/../hybridauth/src/Provider/Disqus.php',
    'Hybridauth\Provider\Dribbble'                                => __DIR__ . '/../hybridauth/src/Provider/Dribbble.php',
    'Hybridauth\Provider\Facebook'                                => __DIR__ . '/../hybridauth/src/Provider/Facebook.php',
    'Hybridauth\Provider\Foursquare'                              => __DIR__ . '/../hybridauth/src/Provider/Foursquare.php',
    'Hybridauth\Provider\Freeagent'                               => __DIR__ . '/../hybridauth/src/Provider/Freeagent.php',
    'Hybridauth\Provider\GitHub'                                  => __DIR__ . '/../hybridauth/src/Provider/GitHub.php',
    'Hybridauth\Provider\Google'                                  => __DIR__ . '/../hybridauth/src/Provider/Google.php',
    'Hybridauth\Provider\Instagram'                               => __DIR__ . '/../hybridauth/src/Provider/Instagram.php',
    'Hybridauth\Provider\Mailru'                                  => __DIR__ . '/../hybridauth/src/Provider/Mailru.php',
    'Hybridauth\Provider\Odnoklassniki'                           => __DIR__ . '/../hybridauth/src/Provider/Odnoklassniki.php',
    'Hybridauth\Provider\OpenID'                                  => __DIR__ . '/../hybridauth/src/Provider/OpenID.php',
    'Hybridauth\Provider\PaypalOpenID'                            => __DIR__ . '/../hybridauth/src/Provider/PaypalOpenID.php',
    'Hybridauth\Provider\PixelPin'                                => __DIR__ . '/../hybridauth/src/Provider/PixelPin.php',
    'Hybridauth\Provider\px500'                                   => __DIR__ . '/../hybridauth/src/Provider/px500.php',
    'Hybridauth\Provider\Reddit'                                  => __DIR__ . '/../hybridauth/src/Provider/Reddit.php',
    'Hybridauth\Provider\Stackoverflow'                           => __DIR__ . '/../hybridauth/src/Provider/Stackoverflow.php',
    'Hybridauth\Provider\Steam'                                   => __DIR__ . '/../hybridauth/src/Provider/Steam.php',
    'Hybridauth\Provider\Tumblr'                                  => __DIR__ . '/../hybridauth/src/Provider/Tumblr.php',
    'Hybridauth\Provider\TwitchTV'                                => __DIR__ . '/../hybridauth/src/Provider/TwitchTV.php',
    'Hybridauth\Provider\Twitter'                                 => __DIR__ . '/../hybridauth/src/Provider/Twitter.php',
    'Hybridauth\Provider\Vkontakte'                               => __DIR__ . '/../hybridauth/src/Provider/Vkontakte.php',
    'Hybrid_Providers_Deezer'                                     => __DIR__ . '/../hybridauth/src/Provider/WIP/Deezer.php',
    'Hybrid_Providers_Goodreads'                                  => __DIR__ . '/../hybridauth/src/Provider/WIP/Goodreads.php',
    'Hybrid_Providers_LinkedIn'                                   => __DIR__ . '/../hybridauth/src/Provider/WIP/LinkedIn.php',
    'Hybrid_Providers_Live'                                       => __DIR__ . '/../hybridauth/src/Provider/WIP/Live.php',
    'Hybrid_Providers_Murmur'                                     => __DIR__ . '/../hybridauth/src/Provider/WIP/Murmur.php',
    'Hybrid_Providers_Odnoklassniki'                              => __DIR__ . '/../hybridauth/src/Provider/WIP/Odnoklassniki.php',
    'Hybrid_Providers_Pixnet'                                     => __DIR__ . '/../hybridauth/src/Provider/WIP/Pixnet.php',
    'Hybrid_Providers_Plurk'                                      => __DIR__ . '/../hybridauth/src/Provider/WIP/Plurk.php',
    'Hybrid_Providers_QQ'                                         => __DIR__ . '/../hybridauth/src/Provider/WIP/QQ.php',
    'Hybrid_Providers_Skyrock'                                    => __DIR__ . '/../hybridauth/src/Provider/WIP/Skyrock.php',
    'Hybrid_Providers_XING'                                       => __DIR__ . '/../hybridauth/src/Provider/WIP/XING.php',
    'Hybrid_Providers_Yahoo'                                      => __DIR__ . '/../hybridauth/src/Provider/WIP/Yahoo.php',
    'Hybrid_Providers_Yandex'                                     => __DIR__ . '/../hybridauth/src/Provider/WIP/Yandex.php',
    'Hybridauth\Provider\WordPress'                               => __DIR__ . '/../hybridauth/src/Provider/WordPress.php',
    'Hybridauth\Provider\YahooOpenID'                             => __DIR__ . '/../hybridauth/src/Provider/YahooOpenID.php',
    'Hybridauth\Storage\Session'                                  => __DIR__ . '/../hybridauth/src/Storage/Session.php',
    'Hybridauth\Storage\StorageInterface'                         => __DIR__ . '/../hybridauth/src/Storage/StorageInterface.php',
    'Hybridauth\Thirdparty\OAuth\OAuthConsumer'                   => __DIR__ . '/../hybridauth/src/Thirdparty/OAuth/OAuthConsumer.php',
    'Hybridauth\Thirdparty\OAuth\OAuthRequest'                    => __DIR__ . '/../hybridauth/src/Thirdparty/OAuth/OAuthRequest.php',
    'Hybridauth\Thirdparty\OAuth\OAuthSignatureMethod'            => __DIR__ . '/../hybridauth/src/Thirdparty/OAuth/OAuthSignatureMethod.php',
    'Hybridauth\Thirdparty\OAuth\OAuthSignatureMethodHMACSHA1'    => __DIR__ . '/../hybridauth/src/Thirdparty/OAuth/OAuthSignatureMethodHMACSHA1.php',
    'Hybridauth\Thirdparty\OAuth\OAuthUtil'                       => __DIR__ . '/../hybridauth/src/Thirdparty/OAuth/OAuthUtil.php',
    'Hybridauth\Thirdparty\OpenID\LightOpenID'                    => __DIR__ . '/../hybridauth/src/Thirdparty/OpenID/LightOpenID.php',
    'Hybridauth\User\Activity'                                    => __DIR__ . '/../hybridauth/src/User/Activity.php',
    'Hybridauth\User\Contact'                                     => __DIR__ . '/../hybridauth/src/User/Contact.php',
    'Hybridauth\User\Profile'                                     => __DIR__ . '/../hybridauth/src/User/Profile.php',
	
	
    'OrgHeiglHybridAuth\Module'                                   => __DIR__ . '/Module.php',
    'OrgHeiglHybridAuth\Controller\IndexController'               => __DIR__ . '/src/OrgHeiglHybridAuth/Controller/IndexController.php',
    'OrgHeiglHybridAuth\HybridAuthUserWrapper'                    => __DIR__ . '/src/OrgHeiglHybridAuth/HybridAuthUserWrapper.php',
    'OrgHeiglHybridAuth\Service\HybridAuthFactory'                => __DIR__ . '/src/OrgHeiglHybridAuth/Service/HybridAuthFactory.php',
    'OrgHeiglHybridAuth\Service\IndexControllerFactory'           => __DIR__ . '/src/OrgHeiglHybridAuth/Service/IndexControllerFactory.php',
    'OrgHeiglHybridAuth\Service\SessionFactory'                   => __DIR__ . '/src/OrgHeiglHybridAuth/Service/SessionFactory.php',
    'OrgHeiglHybridAuth\UserInterface'                            => __DIR__ . '/src/OrgHeiglHybridAuth/UserInterface.php',
    'OrgHeiglHybridAuth\UserWrapperFactory'                       => __DIR__ . '/src/OrgHeiglHybridAuth/UserWrapperFactory.php',
    'OrgHeiglHybridAuth\View\Helper\HybridAuth'                   => __DIR__ . '/src/OrgHeiglHybridAuth/View/Helper/HybridAuth.php',
    'OrgHeiglHybridAuthTest\Bootstrap'                            => __DIR__ . '/tests/Bootstrap.php',
    'OrgHeiglHybridAuthTest\Framework\TestCase'                   => __DIR__ . '/tests/OrgHeiglHybridAuthTest/Framework/TestCase.php',
    'OrgHeiglHybridAuthTest\HybridAuthUserWrapperTest'            => __DIR__ . '/tests/OrgHeiglHybridAuthTest/HybridAuthUserWrapperTest.php',
    'OrgHeiglHybridAuthTest\HybridAuthFactoryTest'                => __DIR__ . '/tests/OrgHeiglHybridAuthTest/Service/HybridAuthFactoryTest.php',
    'OrgHeiglHybridAuthTest\IndexControllerFactoryTest'           => __DIR__ . '/tests/OrgHeiglHybridAuthTest/Service/IndexControllerFactoryTest.php',
    'OrgHeiglHybridAuthTest\SessionFactoryTest'                   => __DIR__ . '/tests/OrgHeiglHybridAuthTest/Service/SessionFactoryTest.php',
    'OrgHeiglHybridAuthTest\UserProxyFactoryTest'                 => __DIR__ . '/tests/OrgHeiglHybridAuthTest/UserWrapperFactoryTest.php',
);

```

**Comment by heiglandreas** [comment]
You're very brave to still try to get it to work. I have an appointment right now but I'll dig into it afterwards in about two to three hours! You are definitely a step further and we'll get the last bit to work also!

**Comment by heiglandreas** [comment]
Seems like I have to apologize for my comment four days ago. There definitely is no HybridAuth-class in the OrgHeiglHybridAuth-Namespace. It has to be the one from the HybridAuth-Namespace.

Instead you can use ```$this->getServiceLocator()->get('OrgHeiglHybridAuthToken')`` in your Controller to get an instance of the ```\OrgHeiglHybridAuth\UserToken```-class that allows you to get the username etc.

Have a look at my implementation on the [php.ug](http://php.ug)-website at https://github.com/php-ug/php.ug/blob/master/src/module/Phpug/Module.php#L98 which adds the user-token to an AccessControllList.

Hope that helps more than my last comments :-/

**Comment by heiglandreas** [comment]
And thanks for the improved script! Great work!

**Comment by lorezyra** [comment]
Thanks for the quick response! I'm trying to get this to work for the top four major Identity providers (in the States & Japan): FaceBook, Twitter, G+, LinkedIn. Starting with FB...

I didn't find the ```OrgHeiglHybridAuthToken``` object anywhere either. I have basically read through both libraries top-to-bottom a couple of times now. Based on what I found in your test classes and view-helper, I decided to build something different for a test page. Here is what I have for my test IndexController.php:

```PHP
/**
 * Zend Framework (http://framework.zend.com/)
 *
 * @link      http://github.com/zendframework/ZendSkeletonApplication for the canonical source repository
 * @copyright Copyright (c) 2005-2013 Zend Technologies USA Inc. (http://www.zend.com)
 * @license   http://framework.zend.com/license/new-bsd New BSD License
 
*
* @author: Rich@RichieBartlett.com
 */

namespace Common\Mvc\Controller;

use Common\Mvc\Controller\BaseController;
use Zend\View\Model\ViewModel;
//use \OrgHeiglHybridAuth\Service\SessionFactory;
use \OrgHeiglHybridAuth\Service\HybridAuthFactory;
use \OrgHeiglHybridAuth\UserInterface;
use \OrgHeiglHybridAuth\UserWrapperFactory;
use \Hybridauth\Hybridauth;
use \Hybridauth\User\Profile;

class IndexController extends BaseController
{
	public function indexAction() {
		$viewData = array();
		
		if($this->isUserAuthenticated()){
			return $this->redirect()->toRoute('feed');
		}




if ( !class_exists('Hybridauth\Hybridauth') ) {
    echo "Could not find Hybridauth class!\n";
} else {
if ( !class_exists('OrgHeiglHybridAuth\HybridAuthUserWrapper') ) {
    echo "Could not find OrgHeiglHybridAuth class!\n";
} else {


	$SL = $this->getServiceLocator();


	$factory = new HybridAuthFactory();
	$hybridAuth = $factory->createService( $SL ); //returns hybridAuth


	$OrgHeiglHybridAuthSession = $SL->get('OrgHeiglHybridAuthSession');


//	\OrgHeiglHybridAuth\UserInterface $userObject = $OrgHeiglHybridAuthSession->container->offsetGet('user');
//	$HAuser = new  \OrgHeiglHybridAuth\UserWrapperFactory();
//	$ssoUser = $HAuser->factory($userObject);
	$proxy = new \OrgHeiglHybridAuth\HybridAuthUserWrapper();
	$userObj = new Profile();



echo '<pre>';
print_r($hybridAuth);
echo "<HR>";
print_r($OrgHeiglHybridAuthSession);
echo "<HR>";
print_r($_SESSION['HA::STORE']);
echo "<HR>";
print_r($userObj);
echo "<HR>";
//print_r($proxy->user);
echo '</pre>';
		

//		$viewData['config'] = $config;
//		$viewData['hybridAuth'] = $hybridAuth;
		$viewData['container'] = $OrgHeiglHybridAuthSession;
		return new ViewModel($viewData);
	}
    
}
}


}

```

Better test page:
```PHP

<!-- TopHeaderArea -->
<!-- //TopHeaderArea -->

<!-- TopMainArea -->
<DIV id="MVC_viewPort"></DIV>
<!-- //TopMainArea -->

<?PHP 

//var_dump($this->container);

if ( $this->container !== NULL) {
	if ( $this->container->offsetExists('authenticated')) {
	
		/** @var OrgHeiglHybridAuth\UserInterface $user */
		$user = $this->container->offsetGet('user');
		var_dump($user);
	
		if( $user !== NULL) {
			echo $user->getName(); // The name of the logged in user
			echo $user->getUID();  // The internal UID of the used service
			echo $user->getMail(); // The mail-address the service provides
			echo $user->getLanguage(); // The language the service provides for the user
		} else {
			echo "<BR><< No user data found... >>\n<BR>";
		}

		$service = $container->offsetGet('backend');
		if( $service !== NULL) {
			echo $service->id; // Should print out the Name of the service provider.
		} else {
			echo "<BR><< No service provider ID found... >>\n<BR>";
		}
} else {
		echo 'No SSO data...';
	}
}
?>
```


Now, progress??
print_r($_SESSION['HA::STORE']);
```
Array
(
    [hauth_session.facebook.hauth_return_to] => s:56:"http://linspira.com/authenticate/login/Facebook/aG9tZQ==";
    [hauth_session.facebook.hauth_endpoint] => s:60:"http://linspira.com/authenticate/backend?hauth.done=Facebook";
    [hauth_session.facebook.id_provider_params] => a:3:{s:15:"hauth_return_to";s:56:"http://linspira.com/authenticate/login/Facebook/aG9tZQ==";s:14:"hauth_endpoint";s:60:"http://linspira.com/authenticate/backend?hauth.done=Facebook";s:15:"hauth_start_url";s:83:"http://linspira.com/authenticate/backend?hauth.start=Facebook&hauth.time=1430974480";}
    [hauth_session.facebook.tokens] => O:41:"Hybridauth\Adapter\Template\OAuth2\Tokens":4:{s:11:"accessToken";N;s:12:"refreshToken";N;s:20:"accessTokenExpiresIn";N;s:20:"accessTokenExpiresAt";N;}
    [hauth_session.error.status] => i:1;
    [hauth_session.error.message] => s:278:"Authentication failed: Provider returned an invalid access token. HTTP client state: (Uri: https://graph.facebook.com/oauth/access_token. Method: POST. Error: 0. Status: 400. Response: {"error":{"message":"Invalid verification code format.","type":"OAuthException","code":100}})";
    [hauth_session.error.code] => i:5;
    [hauth_session.facebook.token.authorization_state] => s:39:"HA-D9KPG8YB674ZJLWAM3S2HEVN5XTOQUC0FR1I";
)
```

This is my first time trying to get FB login to work. So, I am not an expert here. Despite confirming my FB.id & FB.secret, it doesn't work. But I'm not sure if any recent changes on Facebook's side is responsible for the Authentication failures.

Just an observation, this lib doesn't seem to gracefully handle provider errors (unless I missed that section). In the meantime, I will use https://developers.facebook.com/tools/explorer/ (Graph API Explorer) to troubleshoot further. I have also forked this lib and if it's okay, I'll push some updates after I get it working on my environment.

Would you have any thoughts?

**Comment by heiglandreas** [comment]
First of all: Facebook is a bit picky. I know that they change their API frequently without notice. I'd suggest trying Twitter or Google first. If they work I'd see to facebook.

Second: Which branch of the OrgHiglHybridAuth are you using? As multiple providers don't work with the master branch. You'll have to use the multipleProviders-Branch (which I suppose you are already using, just checking the obvious :wink:). 

Third: There is no ```OrgHeiglHybridAuthToken```-class. It's just an identifier for the ServiceManager that will return an instance of ```OrgHeiglHybridAuth\UserToken```. It calls the ```OrgHeiglHybridAuth\Service\UserFactory::createService()``` which will then return a ```UserToken```.

Error handling should be done in the underlying HybridAuth-lib but you might be right on that point. And **yes** I'd be pleased to receive PullRequests :wink:

**Comment by lorezyra** [comment]
Where is the ```OrgHeiglHybridAuth\UserToken``` class?

**Comment by lorezyra** [comment]
I see. Actually, I was using the master branch. Is there a reason why the multipleProviders-Branch is not apart of master?

**Comment by heiglandreas** [comment]
I'm currently investigating the differences. There are some Backwards-Incompatibilities involved. Therefore people using the dev-master-branch might get into trouble when I just merge the multipleProviders into master. 

I'm not yet sure on how I'd do it. Might become a 2.0...

**Comment by lorezyra** [comment]
2.0, Sounds good to me! Since I'm not concerned about the older version of this lib, it makes since to start with the latest. Worse case, I have to roll my own. But I'm hoping that you can save me the time. BTW, I'm serious about the :beers: :moneybag:. Send me a paypal link and I'll donate to your beer (hacker) fund.

**Comment by heiglandreas** [comment]
Let's first get that stuff working and then we'll see about :beers: :wink:

**Comment by heiglandreas** [comment]
AFAIK the issue was resolved, wasn't it? And there's a 2.0.0 out now (based on the multiProvider-branch) If there's anything I can do reopen the issue!
